File: src/main/java/leets/leenk/domain/birthday/domain/service/BirthdayLettersGetService.java
package leets.leenk.domain.birthday.domain.service;

import leets.leenk.domain.birthday.domain.entity.BirthdayLetter;
import leets.leenk.domain.birthday.domain.entity.BirthdayLetterReadMark;
import leets.leenk.domain.birthday.domain.repository.BirthdayLetterReadMarkRepository;
import leets.leenk.domain.birthday.domain.repository.BirthdayLetterRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class BirthdayLettersGetService {
    private final BirthdayLetterRepository birthdayLetterRepository;
    private final BirthdayLetterReadMarkRepository birthdayLetterReadMarkRepository;

    public List<BirthdayLetter> getMyBirthdayLetters(long receiverId) {
        return birthdayLetterRepository.findAllByReceiverIdOrderByCreateDateDesc(receiverId);
    }

    public long countMyReceivedLetters(long receiverId, LocalDateTime start, LocalDateTime end) {
        return birthdayLetterRepository.countByReceiverIdAndCreateDateBetween(receiverId, start, end);
    }

    public Optional<BirthdayLetterReadMark> getBirthdayLetterReadMark(long receiverId) {
        return birthdayLetterReadMarkRepository.findById(receiverId);
    }

    public Boolean hasNewLetters(long loginUserId, LocalDateTime start, LocalDateTime end, LocalDateTime lastReadAt) {
        return birthdayLetterRepository.checkNewBirthdayLetter(loginUserId, start, end, lastReadAt);
    }
}

File: src/main/kotlin/leets/leenk/domain/birthday/domain/service/BirthdayLettersGetService.kt
package leets.leenk.domain.birthday.domain.service

import leets.leenk.domain.birthday.domain.entity.BirthdayLetter
import leets.leenk.domain.birthday.domain.entity.BirthdayLetterReadMark
import leets.leenk.domain.birthday.domain.repository.BirthdayLetterReadMarkRepository
import leets.leenk.domain.birthday.domain.repository.BirthdayLetterRepository
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import java.time.LocalDateTime

@Service
class BirthdayLettersGetService(
    private val birthdayLetterRepository: BirthdayLetterRepository,
    private val birthdayLetterReadMarkRepository: BirthdayLetterReadMarkRepository,
) {
    fun getMyBirthdayLetters(receiverId: Long): List<BirthdayLetter> =
        birthdayLetterRepository.findAllByReceiverIdOrderByCreateDateDesc(receiverId)

    fun countMyReceivedLetters(
        receiverId: Long,
        start: LocalDateTime,
        end: LocalDateTime,
    ): Long = birthdayLetterRepository.countByReceiverIdAndCreateDateBetween(receiverId, start, end)

    fun getBirthdayLetterReadMark(receiverId: Long): BirthdayLetterReadMark? =
        birthdayLetterReadMarkRepository.findByIdOrNull(receiverId)

    fun hasNewLetters(
        loginUserId: Long,
        start: LocalDateTime,
        end: LocalDateTime,
        lastReadAt: LocalDateTime?,
    ): Boolean = birthdayLetterRepository.checkNewBirthdayLetter(loginUserId, start, end, lastReadAt)
}
